xds/internal/balancer/clusterresolver/config.go

```diff
@@ -19,110 +19,44 @@ package clusterresolver
 
 import (
 	"encoding/json"
-	"fmt"
 
-	"google.golang.org/grpc/balancer"
+	internalserviceconfig "google.golang.org/grpc/internal/serviceconfig"
 	"google.golang.org/grpc/serviceconfig"
+	"google.golang.org/grpc/xds/internal/balancer/clusterresolver/balancerconfig"
 )
 
-// EDSConfig represents the loadBalancingConfig section of the service config
-// for EDS balancers.
-type EDSConfig struct {
-	serviceconfig.LoadBalancingConfig
-	// ChildPolicy represents the load balancing config for the child
-	// policy.
-	ChildPolicy *loadBalancingConfig
-	// FallBackPolicy represents the load balancing config for the
-	// fallback.
-	FallBackPolicy *loadBalancingConfig
-	// ClusterName is the cluster name.
-	ClusterName string
-	// EDSServiceName is the name to use in EDS query. If not set, use
-	// ClusterName.
-	EDSServiceName string
-	// MaxConcurrentRequests is the max number of concurrent request allowed for
-	// this service. If unset, default value 1024 is used.
+// LBConfig is the config for cluster resolver balancer.
+type LBConfig struct {
+	serviceconfig.LoadBalancingConfig `json:"-"`
+	// DiscoveryMechanisms is an ordered list of discovery mechanisms.
 	//
-	// Note that this is not defined in the service config proto. And the reason
-	// is, we are dropping EDS and moving the features into cluster_impl. But in
-	// the mean time, to keep things working, we need to add this field. And it
-	// should be fine to add this extra field here, because EDS is only used in
-	// CDS today, so we have full control.
-	MaxConcurrentRequests *uint32
-	// LRS server to send load reports to.  If not present, load reporting
-	// will be disabled.  If set to the empty string, load reporting will
-	// be sent to the same server that we obtained CDS data from.
-	LrsLoadReportingServerName *string
-}
-
-// edsConfigJSON is the intermediate unmarshal result of EDSConfig. ChildPolicy
-// and Fallbackspolicy are post-processed, and for each, the first installed
-// policy is kept.
-type edsConfigJSON struct {
-	ChildPolicy                []*loadBalancingConfig
-	FallbackPolicy             []*loadBalancingConfig
-	ClusterName                string
-	EDSServiceName             string
-	MaxConcurrentRequests      *uint32
-	LRSLoadReportingServerName *string
-}
-
-// UnmarshalJSON parses the JSON-encoded byte slice in data and stores it in l.
-// When unmarshalling, we iterate through the childPolicy/fallbackPolicy lists
-// and select the first LB policy which has been registered.
-func (l *EDSConfig) UnmarshalJSON(data []byte) error {
-	var configJSON edsConfigJSON
-	if err := json.Unmarshal(data, &configJSON); err != nil {
-		return err
-	}
+	// Must have at least one element. Results from each discovery mechanism are
+	// concatenated together in successive priorities.
+	DiscoveryMechanisms []balancerconfig.DiscoveryMechanism `json:"discoveryMechanisms,omitempty"`
```

Comment:
Are the tags required here?

Reply:
All balancer config fields need the json tags to convert from/to JSON. There's also the undocumented rule of whether to use `discoveryMechanisms` (we pick this because of jsonpb) or `discovery_mechanisms` or `DiscoveryMechanisms`